pp.c

```diff
@@ -7934,64 +7934,168 @@ PP(pp_multiparam)
-
         while(argc) {
             SV **svp;
 
             svp = av_fetch(defav, parami, FALSE); parami++;
-            SV *key = svp ? *svp : &PL_sv_undef;
+            SV *name = svp ? *svp : &PL_sv_undef;
             svp = av_fetch(defav, parami, FALSE); parami++;
             SV *val = svp ? *svp : &PL_sv_undef;
             argc -= 2;
 
-            if (UNLIKELY(SvGMAGICAL(key)))
-                key = sv_mortalcopy(key);
+            STRLEN namelen;
+            const char *namepv = SvPV(name, namelen);
+
+            U32 namehash;
+            PERL_HASH(namehash, namepv, namelen);
+
+            PADOFFSET padix = 0;
+            for(size_t namedix = 0; namedix < n_named; namedix++) {
+                struct op_multiparam_named_aux *named = aux->named + namedix;
+
+                if(named->namehash != namehash || named->namelen != namelen)
+                    continue;
+                if(!strnEQ(named->namepv, namepv, namelen))
+                    continue;
+
+                padix = named->padix;
+                break;
+            }
+
+            if(padix) {
+                SV **padentry = &PAD_SVl(padix);
+                save_clearsv(padentry);
+
+                assert(TAINTING_get || !TAINT_get);
+                if (UNLIKELY(TAINT_get) && !SvTAINTED(val))
+                    TAINT_NOT;
+
+                SvPADSTALE_off(*padentry);
+                SvSetMagicSV(*padentry, val);
+            }
+            else if(slurpypad && slurpy == '@') {
+                AV *av = (AV *)*slurpypad;
 
-            hv_store_ent(hv, key, newSVsv(val), 0);
-            if (UNLIKELY(TAINT_get) && !SvTAINTED(val))
-                TAINT_NOT;
+                assert(TAINTING_get || !TAINT_get);
+                if (UNLIKELY(TAINT_get) && !SvTAINTED(name))
+                    TAINT_NOT;
+                if (UNLIKELY(TAINT_get) && !SvTAINTED(val))
+                    TAINT_NOT;
+
+                av_push(av, newSVsv(name));
+                av_push(av, newSVsv(val));
+            }
+            else if(slurpypad && slurpy == '%') {
+                HV *hv = (HV *)*slurpypad;
+
+                if (UNLIKELY(SvGMAGICAL(name)))
+                    name = sv_mortalcopy(name);
+
+                if (UNLIKELY(TAINT_get) && !SvTAINTED(val))
+                    TAINT_NOT;
+                hv_store_ent(hv, name, newSVsv(val), 0);
+            }
+            else {
+                // TODO: Consider collecting up all the names of unrecognised
+                // in one string
+                croak("Unrecognized named parameter '%s' to subroutine '%" SVf "'",
+                    namepv, S_find_runcv_name());
+            }
+        }
+
+        for(size_t namedix = 0; namedix < n_named; namedix++) {
+            struct op_multiparam_named_aux *named = aux->named + namedix;
+            if(!named->is_required || !SvPADSTALE(PAD_SVl(named->padix)))
+                continue;
+
+            // TODO: Consider collecting up all the names of missing
```

Comment:
Yes please (and unrecognized ones above...)

Reply:
I'll fix those TODOs in a follow-up commit once the initial bit is merged. That's more of a "nice-to-have" feature.